routing/checkpoint_predictor.hpp

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+#include "geometry/point2d.hpp"
+
+#include <cstdint>
+#include <iterator>
+#include <vector>
+
+namespace routing
+{
+/// \returns difference between distance |from|->|between| + |between|->|to| and distance |from|->|to|.
+double CalculateDeltaMeters(m2::PointD const & from, m2::PointD const & to, m2::PointD const & between);
+
+/// \brief finds the best position for |point| between points to minimize the length
+/// of broken line of |points| with inserted |point| between them.
+/// \param points is a sequence points on the map.
+/// \param point is a point to be inserted between |points|.
+/// \returns zero based index of a line segment between two points at |points|.
+size_t PredictPosition(std::vector<m2::PointD>::const_iterator checkpointsBegin,
+                       std::vector<m2::PointD>::const_iterator checkpointsEnd,
+                       m2::PointD const & point);
+
+/// \brief This class is responsable for prediction of the order of intermediate route points.
+/// According to current implementation to find the best place for an adding intermediate points
+/// this class inserts an added intermediate point between points at vector
+/// {<route start>, <former intermediate points>, <route finish>}, calculates the length
+/// of the broken line in meters, and finds the sequence of points which gets the shortest length.
+class CheckpointPredictor final
+{
+  friend class CheckpointPredictorTest;
+
+public:
+  CheckpointPredictor(m2::PointD const & start, m2::PointD const & finish)
+    : m_start(start), m_finish(finish)
+  {
+  }
+
+  /// \brief Calls |fn| for each intermediate points from range [|beginIntermediatePoints|, |endIntermediatePoints|).
+  /// Passes as it's params zero based index of the intermediate point at vector
+  /// [|beginIntermediatePoints|, |endIntermediatePoints|) and its position in intermediate point vector
+  /// after adding |addedPoint|.
+  /// \returns index of |addedPoint| at the vector after it's added.
+  template <typename Fn>
+  size_t ForEachIntermediatePoint(std::vector<m2::PointD>::const_iterator beginIntermediatePoints,
+                                  std::vector<m2::PointD>::const_iterator endIntermediatePoints,
+                                  m2::PointD const & addedPoint, Fn && fn) const
+  {
+    std::vector<m2::PointD> checkpoints;
+    FillCheckpoints(beginIntermediatePoints, endIntermediatePoints, checkpoints);
+    // Note. The line segment index which is return by PredictPosition() method is exactly the same with
+    // min index of intermediate point which changes its order. The thing is the line segment index returned
+    // by PredictPosition() includes line segment from |m_start| to zero intermediate point.
```

Comment:
Yes. But zero is an adjective and a noun. Here I used it as an adjective meaning to prevent confusing with *begin and *(begin + 1) items. What is your opinion? Do you think it's better to write "the first item" but not "the zero item"?